### PR TITLE
NMS-13707: Added docs for Flow Thresholding

### PR DESCRIPTION
**Pull request by christianpape**
* JIRA: https://issues.opennms.org/browse/NMS-13707

Please note: using a resource-filter in a flowApp threshold definition is currently broken (see NMS-13945). So the example configs won't work but will work when NMS-13945 is fixed.

### CODE REVIEW

File: docs/modules/operation/pages/flows/data-collection.adoc [hunk 1]
```diff
@@ -0,0 +1,34 @@
+[[ga-flow-support-data-collection]]
+= Data-collection of Flow applications
```
Comment (Bonrob2) [inline docs/modules/operation/pages/flows/data-collection.adoc:2]:
```suggestion
= Data collection of flow applications
```

File: docs/modules/operation/pages/flows/data-collection.adoc [hunk 2]
```diff
@@ -0,0 +1,34 @@
...
+
+Flows are categorized in applications using the <<flows/classification-engine.adoc#ga-flow-support-classification-engine, Classification Engine>>.
+OpenNMS supports to sum up bytesIn/bytesOut data of flow records based on these Flow applications.
```
Comment (Bonrob2) [inline docs/modules/operation/pages/flows/data-collection.adoc:5]:
```suggestion
OpenNMS supports summing up the bytesIn/bytesOut data of flow records based on these flow applications and the collection of this data.
```

File: docs/modules/operation/pages/flows/data-collection.adoc [hunk 3]
```diff
@@ -0,0 +1,34 @@
...
+OpenNMS supports the data-collection of this data.
```
Comment (Bonrob2) [inline docs/modules/operation/pages/flows/data-collection.adoc:6]:
```suggestion
```

File: docs/modules/operation/pages/flows/data-collection.adoc [hunk 4]
```diff
@@ -0,0 +1,34 @@
...
+This enables a user to profile traffic in a network and to also graph this data.
```
Comment (Bonrob2) [inline docs/modules/operation/pages/flows/data-collection.adoc:7]:
```suggestion
This lets a user profile network traffic and also graph this data.
```

File: docs/modules/operation/pages/flows/data-collection.adoc [hunk 5]
```diff
@@ -0,0 +1,34 @@
...
+
+== Configure data-collection
```
Comment (Bonrob2) [inline docs/modules/operation/pages/flows/data-collection.adoc:9]:
```suggestion
== Configure data collection
```

File: docs/modules/operation/pages/flows/data-collection.adoc [hunk 6]
```diff
@@ -0,0 +1,34 @@
...
+
+In order to enable data-collection for Flows you need to enable the flag `applicationDataCollection` in the adapter's definition.
```
Comment (Bonrob2) [inline docs/modules/operation/pages/flows/data-collection.adoc:11]:
```suggestion
To enable data collection for flows you need to enable the `applicationDataCollection` flag in the adapter's definition.
```

File: docs/modules/operation/pages/flows/data-collection.adoc [hunk 7]
```diff
@@ -0,0 +1,34 @@
...
+Furthermore you have to assure that a valid package definition exists inside the adapter's definition.
```
Comment (Bonrob2) [inline docs/modules/operation/pages/flows/data-collection.adoc:12]:
```suggestion
Furthermore, you have to ensure that a valid package definition exists inside the adapter's definition.
```

File: docs/modules/operation/pages/flows/data-collection.adoc [hunk 8]
```diff
@@ -0,0 +1,34 @@
...
+
+.Example of enabling data-collection for the NetFlow v9 adapter in telemetryd-configuration.xml
```
Comment (Bonrob2) [inline docs/modules/operation/pages/flows/data-collection.adoc:14]:
```suggestion
.Example of enabling data collection for the NetFlow v9 adapter in telemetryd-configuration.xml
```

File: docs/modules/operation/pages/flows/thresholding.adoc [hunk 9]
```diff
@@ -0,0 +1,74 @@
+[[ga-flow-support-data-collection]]
+= Thresholding Flow applications
```
Comment (Bonrob2) [inline docs/modules/operation/pages/flows/thresholding.adoc:2]:
```suggestion
= Thresholding flow applications
```

File: docs/modules/operation/pages/flows/thresholding.adoc [hunk 10]
```diff
@@ -0,0 +1,74 @@
...
+
+Flows are categorized in applications using the <<flows/classification-engine.adoc#ga-flow-support-classification-engine, Classification Engine>>.
+OpenNMS supports to sum up bytesIn/bytesOut data of flow records based on these Flow applications.
```
Comment (Bonrob2) [inline docs/modules/operation/pages/flows/thresholding.adoc:5]: OpenNMS supports summing up the bytesIn/bytesOut data of flow records based on these flow applications.

File: docs/modules/operation/pages/flows/thresholding.adoc [hunk 11]
```diff
@@ -0,0 +1,74 @@
...
+Users can define thresholds on this data to generate alarms in the case the amount of traffic violates these thresholds.
```
Comment (Bonrob2) [inline docs/modules/operation/pages/flows/thresholding.adoc:6]:
```suggestion
Users can define thresholds on this data to generate alarms when the amount of traffic violates these thresholds.
```

File: docs/modules/operation/pages/flows/thresholding.adoc [hunk 12]
```diff
@@ -0,0 +1,74 @@
...
+
+== Configure thresholding for Flow applications
```
Comment (Bonrob2) [inline docs/modules/operation/pages/flows/thresholding.adoc:8]:
```suggestion
== Configure thresholding for flow applications
```

File: docs/modules/operation/pages/flows/thresholding.adoc [hunk 13]
```diff
@@ -0,0 +1,74 @@
...
+
+In order to enable thresholding for Flows you need to enable the flag `applicationThresholding` in the adapter's definition.
```
Comment (Bonrob2) [inline docs/modules/operation/pages/flows/thresholding.adoc:10]:
```suggestion
To enable thresholding for flows you need to enable the `applicationThresholding` flag in the adapter's definition.
```

File: docs/modules/operation/pages/flows/thresholding.adoc [hunk 14]
```diff
@@ -0,0 +1,74 @@
...
+        <include-range begin="::1" end="ffff:ffff:ffff:ffff:ffff:ffff:ffff:ffff"/>
+        <service name="Flow-Threshold" interval="300000" user-defined="false" status="on">
+            <parameter key="thresholding-group" value="flow-thresholding-group"/>
+        </service>
+    </package>
+...
+</threshd-configuration>
+----
+
+Finally, you define the thresholds in the `thresholds.xml`.
```
Comment (Bonrob2) [inline docs/modules/operation/pages/flows/thresholding.adoc:54]:
```suggestion
Finally, define the thresholds in the `thresholds.xml` file.
```

File: docs/modules/operation/pages/flows/thresholding.adoc [hunk 15]
```diff
@@ -0,0 +1,74 @@
...
+Use need to use `flowApp` for the `ds-type`.
```
Comment (Bonrob2) [inline docs/modules/operation/pages/flows/thresholding.adoc:55]:
```suggestion
Use `flowApp` for the `ds-type`.
```

File: docs/modules/reference/pages/telemetryd/protocols/ipfix.adoc [hunk 16]
```diff
@@ -60,6 +60,12 @@ The value gives the size of the history buffer allowing flows to be processed ou
 | Fallback value for sampling interval, if value is not included in exported flows.
 | none
 |===
+NOTE: The parameter `maxClockSkew` in your parser definition enables clock skew detection for exporters.
+It specifies the maximum delta in seconds between exporter and Minion timestamps.
+If exceeded, an alarm will be generated for the exporting device.
+The default value is 0, so clock skew detection is disabled.
+Furthermore, you can use a parameter `clockSkewEventRate` to rate limit clock skew events.
```
Comment (Bonrob2) [inline docs/modules/reference/pages/telemetryd/protocols/ipfix.adoc:67]:
```suggestion
Furthermore, you can use the `clockSkewEventRate` parameter to rate limit clock skew events.
```

File: docs/modules/reference/pages/telemetryd/protocols/ipfix.adoc [hunk 17]
```diff
@@ -75,7 +81,48 @@ The IPFIX TCP parser accepts packets received by a <<telemetryd/listeners/tcp.ad
...
+| Parameter
+| Description
+| Default
+
+| maxClockSkew
+| The maximum delta in seconds between exporter and Minion timestamps.
+| 0
+
+| clockSkewEventRate
+| Used to rate-limit clock skew events in seconds.
```
Comment (Bonrob2) [inline docs/modules/reference/pages/telemetryd/protocols/ipfix.adoc:96]:
```suggestion
| Rate limits clock skew events, in seconds.
```

File: docs/modules/reference/pages/telemetryd/protocols/ipfix.adoc [hunk 18]
```diff
@@ -75,7 +81,48 @@ The IPFIX TCP parser accepts packets received by a <<telemetryd/listeners/tcp.ad
...
+| 3600
+
+| dnsLookupsEnabled
+| Used to enable or disable DNS resolution for flows.
```
Comment (Bonrob2) [inline docs/modules/reference/pages/telemetryd/protocols/ipfix.adoc:100]:
```suggestion
| Enables or disables DNS resolution for flows.
```

File: docs/modules/reference/pages/telemetryd/protocols/ipfix.adoc [hunk 19]
```diff
@@ -75,7 +81,48 @@ The IPFIX TCP parser accepts packets received by a <<telemetryd/listeners/tcp.ad
...
+| true
+
+| sequenceNumberPatience
+| A value > 1 enables checking for sequence number completeness.
+The value gives the size of the history buffer allowing flows to be processed out of order.
```
Comment (Bonrob2) [inline docs/modules/reference/pages/telemetryd/protocols/ipfix.adoc:105]: Is it the history buffer allowing flows to be processed out of order, or can someone use the value to allow flows to be processed out of order?

File: docs/modules/reference/pages/telemetryd/protocols/ipfix.adoc [hunk 20]
```diff
@@ -75,7 +81,48 @@ The IPFIX TCP parser accepts packets received by a <<telemetryd/listeners/tcp.ad
...
+
+| flowSamplingIntervalFallback
+| Fallback value for sampling interval, if value is not included in exported flows.
+| none
+|===
+NOTE: The parameter `maxClockSkew` in your parser definition enables clock skew detection for exporters.
+It specifies the maximum delta in seconds between exporter and Minion timestamps.
+If exceeded, an alarm will be generated for the exporting device.
+The default value is 0, so clock skew detection is disabled.
+Furthermore, you can use a parameter `clockSkewEventRate` to rate limit clock skew events.
```
Comment (Bonrob2) [inline docs/modules/reference/pages/telemetryd/protocols/ipfix.adoc:124]:
```suggestion
Furthermore, you can use the `clockSkewEventRate`  parameter to rate limit clock skew events.
```

File: docs/modules/reference/pages/telemetryd/protocols/ipfix.adoc [hunk 21]
```diff
@@ -112,33 +159,19 @@ Received flows are decoded from the messages into the canonical flow format and
...
 |===
 | Parameter
 | Description
 | Default
 
-| templateTimeout
-| Templates must be redeclared in the given duration or they will be dropped.
-| 30 minutes
+| applicationDataCollection
+| Used to enable data-collection of bytesIn/bytesOut based on flow applications.
```
Comment (Bonrob2) [inline docs/modules/reference/pages/telemetryd/protocols/ipfix.adoc:170]:
```suggestion
| Enables data-collection of bytesIn/bytesOut based on flow applications.
```

File: docs/modules/reference/pages/telemetryd/protocols/ipfix.adoc [hunk 22]
```diff
@@ -112,33 +159,19 @@ Received flows are decoded from the messages into the canonical flow format and
...
-
-| clockSkewEventRate
-| Rate-limit clock skew events in seconds.
-| 3600
-
-| dnsLookupsEnabled
-| Enable or disable DNS resolution for flows.
-| true
+| applicationThresholding
+| Used to enable thresholding based on flow applications.
```
Comment (Bonrob2) [inline docs/modules/reference/pages/telemetryd/protocols/ipfix.adoc:174]:
```suggestion
| Enables thresholding based on flow applications.
```

File: docs/modules/reference/pages/telemetryd/protocols/ipfix.adoc [hunk 23]
```diff
@@ -112,33 +159,19 @@ Received flows are decoded from the messages into the canonical flow format and
...
+| false
 |===
-
-NOTE: The parameter `maxClockSkew` in your parser definition enables clock skew detection for exporters.
-It specifies the maximum delta in seconds between exporter and Minion timestamps.
-If exceeded, an alarm will be generated for the exporting device.
-The default value is 0, so clock skew detection is disabled.
-Furthermore, you can use a parameter `clockSkewEventRate` to rate limit clock skew events.
-The default is `3600` seconds, so every hour an event will be sent.
+NOTE: In order the data-collection to work properly you must also define a valid package definition inside your adapter configuration.
```
Comment (Bonrob2) [inline docs/modules/reference/pages/telemetryd/protocols/ipfix.adoc:177]:
```suggestion
NOTE: For data collection to work properly you must also define a valid package definition inside your adapter configuration.
```

File: docs/modules/reference/pages/telemetryd/protocols/netflow5.adoc [hunk 24]
```diff
@@ -62,4 +62,19 @@ Flows are decoded from the messages into the canonical flow format and published
...
-This adapter does not currently have any configurable parameters.
+.Required adapter-specific parameters
+[options="header", cols="1,3,1"]
+|===
+| Parameter
+| Description
+| Default
+
+| applicationDataCollection
+| Used to enable data-collection of bytesIn/bytesOut based on flow applications.
```
Comment (Bonrob2) [inline docs/modules/reference/pages/telemetryd/protocols/netflow5.adoc:73]:
```suggestion
| Enables data collection of bytesIn/bytesOut based on flow applications.
```

File: docs/modules/reference/pages/telemetryd/protocols/netflow5.adoc [hunk 25]
```diff
@@ -62,4 +62,19 @@ Flows are decoded from the messages into the canonical flow format and published
...
+| false
+
+| applicationThresholding
+| Used to enable thresholding based on flow applications.
```
Comment (Bonrob2) [inline docs/modules/reference/pages/telemetryd/protocols/netflow5.adoc:77]:
```suggestion
| Enables thresholding based on flow applications.
```

File: docs/modules/reference/pages/telemetryd/protocols/netflow5.adoc [hunk 26]
```diff
@@ -62,4 +62,19 @@ Flows are decoded from the messages into the canonical flow format and published
...
+| false
+|===
+NOTE: In order the data-collection to work properly you must also define a valid package definition inside your adapter configuration.
```
Comment (Bonrob2) [inline docs/modules/reference/pages/telemetryd/protocols/netflow5.adoc:80]:
```suggestion
NOTE: For data collection to work properly you must also define a valid package definition inside your adapter configuration.
```

File: docs/modules/reference/pages/telemetryd/protocols/netflow9.adoc [hunk 27]
```diff
@@ -84,4 +84,19 @@ Flows are decoded from the messages into the canonical flow format and are publi
...
-This adapter does not currently have any configurable parameters.
+.Required adapter-specific parameters
+[options="header", cols="1,3,1"]
+|===
+| Parameter
+| Description
+| Default
+
+| applicationDataCollection
+| Used to enable data-collection of bytesIn/bytesOut based on flow applications.
```
Comment (Bonrob2) [inline docs/modules/reference/pages/telemetryd/protocols/netflow9.adoc:95]:
```suggestion
| Enables data collection of bytesIn/bytesOut based on flow applications.
```

File: docs/modules/reference/pages/telemetryd/protocols/netflow9.adoc [hunk 28]
```diff
@@ -84,4 +84,19 @@ Flows are decoded from the messages into the canonical flow format and are publi
...
+| false
+
+| applicationThresholding
+| Used to enable thresholding based on flow applications.
```
Comment (Bonrob2) [inline docs/modules/reference/pages/telemetryd/protocols/netflow9.adoc:99]:
```suggestion
| Enables thresholding based on flow applications.
```

File: docs/modules/reference/pages/telemetryd/protocols/netflow9.adoc [hunk 29]
```diff
@@ -84,4 +84,19 @@ Flows are decoded from the messages into the canonical flow format and are publi
...
+| false
+|===
+NOTE: In order the data-collection to work properly you must also define a valid package definition inside your adapter configuration.
```
Comment (Bonrob2) [inline docs/modules/reference/pages/telemetryd/protocols/netflow9.adoc:102]:
```suggestion
NOTE: For data collection to work properly you must also define a valid package definition inside your adapter configuration.
```

File: docs/modules/reference/pages/telemetryd/protocols/sflow.adoc [hunk 30]
```diff
@@ -79,7 +79,16 @@ Using the script extension, you can extract the desired metrics from the sFlow m
 | script
 | Full path to the script used to handle the sFlow messages.
 | none
+
+| applicationDataCollection
+| Used to enable data-collection of bytesIn/bytesOut based on flow applications.
```
Comment (Bonrob2) [inline docs/modules/reference/pages/telemetryd/protocols/sflow.adoc:84]:
```suggestion
| Enables data collection of bytesIn/bytesOut based on flow applications.
```

File: docs/modules/reference/pages/telemetryd/protocols/sflow.adoc [hunk 31]
```diff
@@ -79,7 +79,16 @@ Using the script extension, you can extract the desired metrics from the sFlow m
...
+| false
+
+| applicationThresholding
+| Used to enable thresholding based on flow applications.
```
Comment (Bonrob2) [inline docs/modules/reference/pages/telemetryd/protocols/sflow.adoc:88]:
```suggestion
| Enables thresholding based on flow applications.
```

File: docs/modules/reference/pages/telemetryd/protocols/sflow.adoc [hunk 32]
```diff
@@ -79,7 +79,16 @@ Using the script extension, you can extract the desired metrics from the sFlow m
...
+| false
 |===
+NOTE: In order the data-collection to work properly you must also define a valid package definition inside your adapter configuration.
```
Comment (Bonrob2) [inline docs/modules/reference/pages/telemetryd/protocols/sflow.adoc:91]:
```suggestion
NOTE: For data collection to work properly you must also define a valid package definition inside your adapter configuration.
```

**Review by Bonrob2 — CHANGES_REQUESTED** [review]
Looking good. A few minor changes.

File: docs/modules/reference/pages/telemetryd/protocols/ipfix.adoc (hunk 19 above)
Comment (christianpape) [reply docs/modules/reference/pages/telemetryd/protocols/ipfix.adoc:105]: The history buffer allows to process the flows out of order.

**Review by Bonrob2 — APPROVED** [review]
LGTM!

File: docs/modules/operation/pages/flows/thresholding.adoc [hunk 33]
```diff
@@ -0,0 +1,74 @@
+[[ga-flow-support-data-collection]]
+= Thresholding flow applications
+
+Flows are categorized in applications using the <<flows/classification-engine.adoc#ga-flow-support-classification-engine, Classification Engine>>.
+OpenNMS supports summing up the bytesIn/bytesOut data of flow records based on these flow applications.
```
[the remark this answers is not in the document]
Comment (indigo423) [reply docs/modules/operation/pages/flows/thresholding.adoc:5]: The overview can explain a bit more what someone needs to do to enable application thresholding. Is application data collection required or does thresholding works also without the data collection?

File: docs/modules/operation/pages/flows/data-collection.adoc [hunk 34]
```diff
@@ -0,0 +1,33 @@
+[[ga-flow-support-data-collection]]
+= Data collection of flow applications
+
+Flows are categorized in applications using the <<flows/classification-engine.adoc#ga-flow-support-classification-engine, Classification Engine>>.
+OpenNMS supports summing up the bytesIn/bytesOut data of flow records based on these flow applications and the collection of this data.
+This lets a user profile network traffic and also graph this data.
```
[the remark this answers is not in the document]
Comment (christianpape) [reply docs/modules/operation/pages/flows/data-collection.adoc:6]: done.

File: docs/modules/operation/pages/flows/data-collection.adoc [hunk 35]
```diff
@@ -0,0 +1,33 @@
...
+
+== Configure data collection
+
+To enable data collection for flows you need to enable the `applicationDataCollection` flag in the adapter's definition.
+Furthermore, you have to ensure that a valid package definition exists inside the adapter's definition.
```
[the remark this answers is not in the document]
Comment (christianpape) [reply docs/modules/operation/pages/flows/data-collection.adoc:11]: done.

File: docs/modules/operation/pages/flows/thresholding.adoc (hunk 33 above)
[the remark this answers is not in the document]
Comment (christianpape) [reply docs/modules/operation/pages/flows/thresholding.adoc:5]: Another level in the hierarchy makes no sense to me. For the other comment, I added sentences to each of the two documents, that data collection does not require thresholding and vice versa.

File: docs/modules/operation/pages/flows/data-collection.adoc [hunk 36]
```diff
@@ -3,12 +3,16 @@
...
+It gives you the possibility of profiling the network throughput by an application.
+
+NOTE: The flow classification engine include a so-called `pre-defined group`, which defines over 6200 applications for basic communication protocols. If you don't want to collect this kind of data you can disable this predefined set of rules in the `Settings` tab on the `Flow Classification` configuration page.
 
 == Configure data collection
 
 To enable data collection for flows you need to enable the `applicationDataCollection` flag in the adapter's definition.
 Furthermore, you have to ensure that a valid package definition exists inside the adapter's definition.
+This may not be necessary when using Newts- or Cortex-based storage.
+Please note, that data collection on flow applications also works without enabling thresholding on flow applications.
```
Comment (Bonrob2) [inline docs/modules/operation/pages/flows/data-collection.adoc:15]:
```suggestion
Note that data collection on flow applications also works without enabling thresholding on flow applications.
```

File: docs/modules/operation/pages/flows/thresholding.adoc [hunk 37]
```diff
@@ -8,6 +8,7 @@ Users can define thresholds on this data to generate alarms when the amount of t
 == Configure thresholding for flow applications
 
 To enable thresholding for flows you need to enable the `applicationThresholding` flag in the adapter's definition.
+Please note, that thresholding on flow applications also works without enabling data collection on flow applications.
```
Comment (Bonrob2) [inline docs/modules/operation/pages/flows/thresholding.adoc:11]:
```suggestion
Note that thresholding on flow applications also works without enabling data collection on flow applications.
```

**Review by Bonrob2 — COMMENTED** [review]
Minor changes.

**Review by Bonrob2 — APPROVED** [review]
LGTM!

**Review by indigo423 — APPROVED** [review]
LGTM!

**Pull request merged**